core/storage/trie_pruner/impl/trie_pruner_impl.hpp

```diff
@@ -0,0 +1,229 @@
+        f(node, count);
+      }
+    }
+
+    std::optional<uint32_t> getPruningDepth() const override {
+      return pruning_depth_;
+    }
+
+   private:
+    TriePrunerImpl(
+        std::optional<uint32_t> pruning_depth,
+        std::shared_ptr<storage::trie::TrieStorageBackend> trie_storage,
+        std::shared_ptr<const storage::trie::TrieSerializer> serializer,
+        std::shared_ptr<const storage::trie::Codec> codec,
+        std::shared_ptr<storage::SpacedStorage> storage,
+        std::shared_ptr<const crypto::Hasher> hasher)
+        : trie_storage_{trie_storage},
+          serializer_{serializer},
+          codec_{codec},
+          storage_{storage},
+          hasher_{hasher},
+          pruning_depth_{pruning_depth} {
+      BOOST_ASSERT(trie_storage_ != nullptr);
+      BOOST_ASSERT(serializer_ != nullptr);
+      BOOST_ASSERT(codec_ != nullptr);
+      BOOST_ASSERT(storage_ != nullptr);
+      BOOST_ASSERT(hasher_ != nullptr);
+    }
+
+    outcome::result<void> prune(primitives::BlockHeader const &state);
+
+    // policy of adding a new trie to the pruner ref counters
+    struct AddConfig {
+      enum AddType {
+        // register only the nodes which are already loaded in memory with the
+        // pruner (these are the nodes that were modified between this trie and
+        // its parent)
+        AddLoadedOnly,
+        // register only the nodes which are altered between this trie and its
+        // parent, loading dummy nodes when required
+        AddNewLoadDummies,
+        // load the whole state into memory and register it with the pruner
+        // (this is a pretty expensive operation)
+        AddWholeState,
+      } type;
+
+      bool shouldLoadDummies() const {
+        switch (type) {
+          case AddLoadedOnly:
+            return false;
+          case AddNewLoadDummies:
+            return true;
+          case AddWholeState:
+            return true;
+        }
+        BOOST_ASSERT_MSG(false, "Unreachable");
+        return false;
+      }
+
+      bool shouldAddAllNodes() const {
+        switch (type) {
+          case AddLoadedOnly:
+            return false;
+          case AddNewLoadDummies:
+            return false;
+          case AddWholeState:
+            return true;
+        }
+        BOOST_ASSERT_MSG(false, "Unreachable");
+        return false;
+      }
+    };
+
+    outcome::result<storage::trie::RootHash> addNewStateWith(
+        trie::PolkadotTrie const &new_trie,
+        trie::StateVersion version,
+        AddConfig config);
+
+    // resets the pruner state and builds it from scratch from the state
+    // of the provided block
+    outcome::result<void> restoreState(
+        primitives::BlockHeader const &base_block,
+        blockchain::BlockTree const &block_tree);
+
+    // store the persistent pruner info to the database
+    outcome::result<void> savePersistentState() const;
+
+    std::unordered_map<common::Buffer, size_t> ref_count_;
```

Comment:
```suggestion
    std::unordered_map<common::Hash256, size_t> ref_count_;
```

Reply:
Merkle value can be shorter than a hash.

Reply:
We count references for "hashes", not "merke values"